### ISSUE TITLE
Bug: no-standalone-expect raises flag says that it raises flag on expect statement that will not be executed. However it also raises flags on expect statements in beforeEach methods that are executed. 

### ISSUE THREAD

**Issue by PramodKumarYadav** [playwright-community/eslint-plugin-playwright#259]
### Issue Summary:

no-standalone-expect says that : `This rule aims to eliminate expect statements that will not be executed. `

However, I added a fixture for asserting network status codes and used a beforeEach method instead of a regular fixture syntax. The `no-standalone-expect `plugin complained that this is outside test and will not be executed but on testing I found that the assert statement was actually executed and thus the warning was misleading. 

### Suggestion:
Since similar to asserting in util functions, there could be items worth asserting in before/after block to check state before moving ahead with tests, we can adjust our plugin rule to not show this error for before/after blocks. 

### Reference
https://github.com/playwright-community/eslint-plugin-playwright/blob/main/docs/rules/no-standalone-expect.md

**Comment by pkumanowski** [comment]
I've also encountered this issue. It happens also when you change name of test block to something different. If you use setup function in pw it will trigger lint error when you have expect in it.

**Comment by mskelton** [comment]
@PramodKumarYadav Can you provide a code example of what you are doing that is being flagged as an error?

**Comment by PramodKumarYadav** [comment]
@mskelton : Sure. Here I have created a file named `verify-status-code.fixture.ts`. Below is the content of this file. 

```
import { test as base, expect } from "@playwright/test";

export const test = base.extend<{ network: void }>({
  network: async ({}, use) => {
    await use();
  },
});

test.beforeEach(async ({ page }) => {
  console.log("Register response event before every test");
  page.on("response", (response) => {
    expect.soft(response.status(), `failed for URL: ${response.url()}`).toBeLessThan(400);
  });
});

test.afterEach(async () => {
  console.log("run after every test");
});
```

When I go to this file, I get this error from the plugin. 

<img width="1114" alt="Screenshot 2024-02-27 at 21 45 24" src="https://github.com/playwright-community/eslint-plugin-playwright/assets/11694972/b3363ccf-5909-45c3-8446-ae7acacde83b">

P.S: To keep things in perspective, I get this error in one project (on Mac machine) but not in another (also on Mac machine) for the exact same configuration in .eslintrc.cjs file - so I found that weird.

Error seen in plugin version:    
 "eslint-plugin-playwright": "^1.3.0", 
I upgraded it to the latest version and saw the same error in one Mac machine. 
"eslint-plugin-playwright": "^1.4.1",

**Comment by mskelton** [comment]
@PramodKumarYadav That code block is a major anti-pattern in Playwright.  Using `page.on` is dangerous as there is zero guarantee that it will be triggered based on the timing of the test.

This is a common mistake when waiting for requests which is why Playwright has `page.waitFor*` methods like this:

```ts
const [request] = await Promise.all([
  page.waitForRequest('https://example.com/resource'),
  page.getByText('trigger request').click()
])
```

Now, all that said, I think I understand what you are trying to do which is mark the test as failed if there are any network errors during the lifetime of the test. This is a pretty niche case so I'll recommend that you simply disable the lint rule for that line. In general, I would recommend against using failed network requests as an indicator of tests passing/failing for several reasons:

- If a network request fails, likely your tests will also fail.
- If a request fails and your tests still pass, it's probably an unrelated network request to the given test.
- Just because a network request fails does not mean your app is not working, it could easily be a network issue and not an application issue.
- If you want to ensure that API requests to a backend are not failing, a better solution is testing the underlying data. For example, if you are testing creating users, after creating a user, make an HTTP request to your backend API with Playwright and verify the user exists. Alternatively use `waitForRequest` when clicking the "Create button" and inspect the response.

In general, it's best to avoid global expectations like this as it gives a false sense of test coverage and also makes it easy for subtle errors to sneak in (if the test finishes before the final network request fails, it will silently pass).

I'm going to close this as not planned.

**Comment by PramodKumarYadav** [comment]
@mskelton : Thanks for all the advice but dont you see the issue with the plugin that it says it will not run a expect statement but it does? 

I believe that is the core of the bug and not if the usage is an anti-pattern or not . Agree?

**Comment by mskelton** [comment]
@PramodKumarYadav I'll update the docs to be more clear, but the behavior of the rule is working as intended.
